pandas/io/parsers/base_parser.py

```diff
@@ -621,12 +621,15 @@ def _check_data_length(
             empty_str_or_na = empty_str | isna(data[-1])  # type: ignore[operator]
             if len(columns) == len(data) - 1 and np.all(empty_str_or_na):
                 return
-            warnings.warn(
-                "Length of header or names does not match length of data. This leads "
-                "to a loss of data with index_col=False.",
-                ParserWarning,
-                stacklevel=find_stack_level(),
-            )
+            # Don't warn if on_bad_lines is set to handle bad lines
+            if self.on_bad_lines == self.BadLineHandleMethod.ERROR:
```

Comment:
This isn't quite clear to me; are you considering `BadLineHandleMethod.WARNING` as being "set to handle bad lines"? Also, why are we warning if it is set to error?